File: src/main/java/io/github/intellij/dlanguage/run/DlangRunDubConfigurationType.java
package io.github.intellij.dlanguage.run;

import com.intellij.execution.configurations.ConfigurationType;
import com.intellij.execution.configurations.ConfigurationTypeBase;
import com.intellij.execution.configurations.RunConfiguration;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.io.FileUtil;
import io.github.intellij.dlanguage.DlangBundle;
import io.github.intellij.dlanguage.icons.DlangIcons;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class DlangRunDubConfigurationType extends ConfigurationTypeBase {

    public DlangRunDubConfigurationType() {
        super("DlangRunDubConfigType",
            DlangBundle.INSTANCE.message("run.dub.text"),
            DlangBundle.INSTANCE.message("run.dub.descr"),
            DlangIcons.RUN);
        addFactory(new DLanguageFactory(this));
    }

    private static class DLanguageFactory extends DlangRunConfigurationFactory {

        public DLanguageFactory(final ConfigurationType type) {
            super(type);
        }

        @NotNull
        public RunConfiguration createTemplateConfiguration(@NotNull final Project project) {
            return new DlangRunDubConfiguration("DlangRunDubConfig", project, this);
        }

        @Override
        public boolean isApplicable(@NotNull Project project) {
            @Nullable final String file = FileUtil.findFileInProvidedPath(project.getBasePath(), "dub.json", "dub.sdl");
            return file != null && !file.isEmpty();
        }
    }
}

File: dub/src/main/kotlin/io/github/intellij/dub/run/DlangRunDubConfigurationType.kt
package io.github.intellij.dub.run

import com.intellij.execution.configurations.ConfigurationType
import com.intellij.execution.configurations.ConfigurationTypeBase
import com.intellij.execution.configurations.RunConfiguration
import com.intellij.openapi.project.Project
import com.intellij.openapi.util.io.FileUtil
import io.github.intellij.dlanguage.DLanguage
import io.github.intellij.dlanguage.DlangBundle.message
import io.github.intellij.dlanguage.run.DlangRunConfigurationFactory

class DlangRunDubConfigurationType : ConfigurationTypeBase(
    "DlangRunDubConfigType",
    message("run.dub.text"),
    message("run.dub.descr"),
    DLanguage.Icons.RUN
) {
    init {
        addFactory(DLanguageFactory(this))
    }

    private class DLanguageFactory(type: ConfigurationType?) : DlangRunConfigurationFactory(type!!) {
        override fun createTemplateConfiguration(project: Project): RunConfiguration {
            return DlangRunDubConfiguration("DlangRunDubConfig", project, this)
        }

        override fun isApplicable(project: Project): Boolean {
            val file = FileUtil.findFileInProvidedPath(project.basePath!!, "dub.json", "dub.sdl")
            return !file.isNullOrEmpty()
        }
    }
}
